test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace TestResultCoordinator.Report.DirectMethodReport
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil.NetworkController;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil.TestResults;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    class NetworkStatusTimeline
+    {
+        static readonly ILogger Logger = ModuleUtil.CreateLogger(nameof(NetworkStatusTimeline));
+
+        readonly IReadOnlyList<NetworkControllerTestResult> networkControllerTestResults;
+        readonly TimeSpan tolerancePeriod;
+
+        public static async Task<NetworkStatusTimeline> Create(ITestResultCollection<TestOperationResult> networkControllerTestOperationResults, TimeSpan tolerancePeriod)
+        {
+            List<NetworkControllerTestResult> networkControllerTestResults = new List<NetworkControllerTestResult>();
+
+            while (await networkControllerTestOperationResults.MoveNextAsync())
+            {
+                Option<NetworkControllerTestResult> networkControllerTestResult = GetNetworkControllerTestOperationResult(networkControllerTestOperationResults.Current);
+
+                networkControllerTestResult.ForEach(
+                    r =>
+                    {
+                        networkControllerTestResults.Add(r);
+                    });
+            }
+
+            return new NetworkStatusTimeline(networkControllerTestResults, tolerancePeriod);
+        }
+
+        NetworkStatusTimeline(IReadOnlyList<NetworkControllerTestResult> networkControllerTestResults, TimeSpan tolerancePeriod)
+        {
+            this.networkControllerTestResults = networkControllerTestResults;
+            this.tolerancePeriod = tolerancePeriod;
+        }
+
+        public NetworkControllerStatus GetNetworkControllerStatusAt(DateTime dateTime)
+        {
+            // Return network controller status at given time
+            NetworkControllerStatus networkControllerStatus = NetworkControllerStatus.Enabled;
+            for (int i = 0;  i < this.networkControllerTestResults.Count; i++)
+            {
+                // If given DateTime is temporally after a DateTime on the list, set the networkStatus to that
+                if (dateTime >= this.networkControllerTestResults[i].CreatedAt)
```

Comment:
First you can't assume given networkControllerTestRsults is sorted.  Second, I think you should try to find the largest created at that is less than given dateTime; and you need to throw exception if given datetime is less than all network controller test result created at timestamp.

Reply:
Also, I think you should combine IsWithinTolerancePeriod with this method and return both NetworkControllerStatus and IsWithinTolerancePeriod boolean value.

Reply:
Combined the two methods and now not assuming order